src/raw_rtde/connection.rs

```diff
@@ -0,0 +1,891 @@
+
+enum RTDECommand {
+    ConfigureInputs = b'I' as isize,
+    ConfigureOutputs = b'O' as isize,
+    Start = b'S' as isize,
+    Pause = b'P' as isize,
+    RtdeDataPacket = b'U' as isize,
+    RequestProtocolVersion = b'V' as isize,
+    GetControllerVersion = b'v' as isize,
+}
+const WOKEN_BEFORE_READ_PSEUDO_CMD: u8 = 255;
+
+/// A low-level class to control RTDE at the level of single packets.
+///
+/// Except for tests it's recommended to use `RawRTDE` to send RTDE over TCP/IP,
+/// as the UR robot expects to receive it.
+pub struct GenericRawRTDE<T: Read + Write + event::Source> {
+    stream: T,
+    poll: Poll,
+    waker: Arc<Waker>,
+    wait_for_reply_timeout: Duration, // Overridden in tests.
+}
+
+/// A raw RTDE connection over TCP/IP.
+pub type RawRTDE = GenericRawRTDE<TcpStream>;
+
+const RTDE_VERSION: u16 = 2;
+const HEADER_SIZE: usize = 3;
+
+const SOCKET_WAIT_TOKEN: Token = Token(0);
+const WAKE_TOKEN: Token = Token(1);
+
+enum ReadResult {
+    ReadData,
+    WokenBeforeRead,
+}
+
+/// Common method of a real RawRTDE and FakeRawRTDE.
+pub trait RawRTDETrait {
+    /// Informs the robot to start sending data. `configure_outputs` need to be called first.
+    fn start_robot_to_host_transmition(&mut self) -> Result<(), String>;
+    /// Asks the robot to stop sending data.
+    fn pause_robot_to_host_transmition(&mut self) -> Result<(), String>;
+    /// Specifies which RTDE variables we want to write to.
+    ///
+    /// There can be several RTDE connections to the robot, but each variable can be
+    /// written to by at most one connection.
+    fn configure_inputs<U: 'static + AsRef<str>>(
+        &mut self,
+        variables: &[U],
+    ) -> Result<IoSetup, String>;
+    /// Specifies which RTDE variables we want to receive.
+    ///
+    /// There can be several RTDE connections to the robot, and each variable can be
+    /// read to by multiple of them.
+    fn configure_outputs<U: 'static + AsRef<str>>(
+        &mut self,
+        frequency: f64,
+        variables: &[U],
+    ) -> Result<IoSetup, String>;
+    /// Sends a packet to the robot. `configure_inputs` needs to be called first.
+    fn send_robot_input(
+        &mut self,
+        io_setup: &IoSetup,
+        input_data: &[RTDEData],
+    ) -> Result<(), String>;
+
+    /// Waits for a packet send by the robot.
+    ///
+    /// One needs to call `configure_outputs` and `start_robot_to_host_transmition` to receive such packets.
+    ///
+    /// May return None is someone used the waker from `get_wait_for_robot_output_waker` to interrupt the call.
+    fn wait_for_robot_output(
+        &mut self,
+        io_setup: &IoSetup,
+    ) -> Result<Option<Vec<RTDEData>>, String>;
+
+    /// Returns a closure that can be called to interrupt a wait_for_robot_output (happening in another thread).
+    fn get_wait_for_robot_output_interrupter(
+        &self,
+    ) -> Box<dyn Fn() -> Result<(), String> + Send + 'static>;
```

Comment:
a `dyn Fn` that's intended to be thread-safe should probably be `Sync` too. (`Sync` is thread safety by-reference, `Send` is thread safety by-value. If you have an `Fn` that can be called multiple times by reference, then chances are `Send` will not be enough. `Send`-only would be warranted for `FnOnce()`.)